Add lootbox tutorial to docs (#1357)
smartcontractkit/documentation: src/pages/vrf/v2/lootbox.mdx

title: "Create a lootbox to distribute random rewards"
---

import { Aside } from "@components"
import { Tabs } from "@components/Tabs"

The lootbox is a utility contract that allow users to open it to receive a random reward.

It supports [ERC-20](https://eips.ethereum.org/EIPS/eip-20), [ERC-721](https://eips.ethereum.org/EIPS/eip-721), and [ERC-1155](https://eips.ethereum.org/EIPS/eip-1155) tokens as rewards. The rewards are distributed from a pool of tokens that are transferred to the contract when the lootbox contract is deployed.

The lootbox contract uses VRF to generate a random number that is used to determine the reward.

<Aside type="caution" title="Disclaimer">
  This repository represents an example of using a Chainlink product or service. It is provided to to help you
  understand how to interact with Chainlink's systems so that you can integrate them into your own. This template is
  provided "AS IS" without warranties of any kind, has not been audited, and may be missing key checks or error handling
  to make the usage of the product more clear. Take everything in this repository as an example and not something to be
  copy pasted into a production ready service.

> </Aside>

Author: I've placed it where I think it looks best, as a placeholder until we decide where it should go.

> </Aside>

Author: Confirmed I have the correct placement and updated it with the latest text.

## Before you begin

Before you start this tutorial, complete the following items:

- Install [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)

>   - To check your installation, run `git --version`. You should see an output similar to `git version x.x.x`.

Author: No, I don't see a minimum git version in the tutorial's [dependencies](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/package-lock.json).

- Install [Nodejs](https://nodejs.org/en/) 16.0.0 or higher
  - To check your installation, run `node --version`. You should see an output similar to `v16.x.x`.

> - Make sure the deployer account has enough LINK to fund the [VRF subscription](/vrf/v2/subscription). If you don't already have a VRF subscription, the deployment script creates one for you.

Reviewer: ```suggestion
- Ensure the deployer account has enough LINK to fund the [VRF subscription](/vrf/v2/subscription). If you don't already have a VRF subscription, the deployment script will create one for you.
```

Reviewer: Shall we also add the link to this page: https://docs.chain.link/resources/fund-your-contract?parent=gettingStarted

  - [Estimate the minimum subscription balance](/vrf/v2/estimating-costs#estimate-minimum-subscription-balance) that VRF requires to process your randomness request. The minimum subscription balance provides a buffer against gas volatility, and only the actual cost of your request will be deducted from your account. If your subscription is underfunded, your VRF request will be pending for 24 hours. If that happens, check the [Subscription Manager](https://vrf.chain.link]) to see the additional balance needed.
  - Configure the initial funding amount in [`network-config.ts`](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/network-config.ts).

>   - For testnets, you can use the [LINK faucet](https://faucets.chain.link/). Request 20 LINK from the faucet - this amount should be sufficient to run this tutorial with a testnet.

Reviewer: ```suggestion
  - To deploy this contract on testnets, use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK. Request 20 LINK from the faucet - this amount should be sufficient to run this tutorial on a testnet.
```

Reviewer: Shall we also add this: https://docs.chain.link/resources/acquire-link?parent=gettingStarted

| `LOOTBOX_AMOUNT_DISTRIBUTED_PER_OPEN` | The amount of reward units distributed per open                                                                           | `1`          |
| `LOOTBOX_OPEN_START_TIMESTAMP`        | The start timestamp in UNIX time for the public open. Leave blank to start immediately.                                   | `1630000000` |
| `VRF_SUBSCRIPTION_ID`                 | A funded Chainlink VRF subscription ID. If you leave this blank, a new subscription will be created and funded on deploy. | `123`        |

#### Reward tokens

For rewards, the lootbox contracts supports any tokens from the following [token standards](https://ethereum.org/en/developers/docs/standards/tokens/):

- [ERC-20](https://ethereum.org/en/developers/docs/standards/tokens/erc-20/): the fungible token standard
- [ERC-721](https://ethereum.org/en/developers/docs/standards/tokens/erc-721/): the NFT standard
- [ERC-1155](https://ethereum.org/en/developers/docs/standards/tokens/erc-1155/): the multi-token standard

> The amounts per unit are used to calculate the lootbox supply of rewards. For example, if you want to distribute 100 tokens of a specific ERC20 token, you would set the `totalAmount` to `100000000000000000000` and the `amountPerUnit` to `1000000000000000000`. This would result in a lootbox supply of 100 units.

Reviewer: ```suggestion
The amounts per unit are used to calculate the lootbox supply of rewards. For example, if you want to distribute 100 tokens of a specific ERC-20 token, you would set the `totalAmount` to `100000000000000000000` and the `amountPerUnit` to `1000000000000000000`. This would result in a lootbox supply of 100 units.
```

  "totalAmount": "100000000000000000000",
  "amountPerUnit": "10000000"
}
```
</Fragment>
<Fragment slot="panel.2">
```json
{
  "tokenType": "ERC721",
  "assetContract": "0x0000000000000000000000000000000000000002",
  "tokenIds": ["1", "2"]
}
```
</Fragment>
<Fragment slot="panel.3">
```json
{
  "tokenType": "ERC1155",
  "assetContract": "0x0000000000000000000000000000000000000003",
  "tokenId": "0",
  "totalAmount": "100",
  "amountPerUnit": "10"
}
```
</Fragment>
</Tabs>

#### Whitelist

The [Merkle tree](https://ethereum.org/en/developers/tutorials/merkle-proofs-for-offline-data-integrity/#introduction) for the private opening stage is generated from the address list in the [`scripts/data/whitelist.json`](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/scripts/data/whitelist.json) file. Edit the file and add all the addresses you want to list.

If you don't want to do a private mint, leave the `whitelist.json` file empty, and the contract will be initialized in [public opening mode](#public-mode).

## Run steps

### Test

- To run the unit tests, run the following command:

  ```bash
  npm run test
  ```

> - If you want to see gas usage, run the following command:

Reviewer: ```suggestion
- To view gas usage, run the following command:
```

  ```bash
  REPORT_GAS=true npm run test
  ```

- For coverage reports, run the following command:

  ```bash
  npm run coverage
  ```

### Deploy

> To run the deploy script, run the following command and replace `<network>` with the network you want to deploy to.

Reviewer: ```suggestion
To run the deploy script, run the following command and replace `<network>` with the network you want to deploy to. The network must be configured in [`hardhat.config.ts`](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/hardhat.config.ts).
```

```bash
npx hardhat run scripts/deploy.js --network <network>
```

> The network must be configured in [`hardhat.config.ts`](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/hardhat.config.ts).

Reviewer: ```suggestion
```

Besides deploying the contract, the [deploy script](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/scripts/deploy.ts) will also:

1.  Approve each token configured in [`scripts/data/tokens.json`](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/scripts/data/tokens.json) for the deployed contract address.

>     This is a mandatory step because the contract must store the tokens in its own balance. So make sure the deployer account owns all the tokens you want to distribute as rewards.

Reviewer: ```suggestion
    This is a mandatory step because the contract must store the tokens in its own balance. It is important to ensure the deployer account owns all the tokens you want to distribute as rewards.
```

1.  Generate a [Merkle tree](https://ethereum.org/en/developers/tutorials/merkle-proofs-for-offline-data-integrity/#introduction) for the whitelist.

1.  Create and fund a VRF subscription if one is not provided.

    {/* prettier-ignore */}
    <Aside type="note">
    Make sure the deployer account has enough LINK to fund the subscription. The initial funding amount is configured in [`network-config.ts`](https://github.com/smartcontractkit/quickstarts-lootbox/blob/main/network-config.ts). For testnets, you can use the [LINK faucet](https://faucets.chain.link/).

    [Estimate the minimum subscription balance](/vrf/v2/estimating-costs#estimate-minimum-subscription-balance) that VRF requires to process your randomness request. The minimum subscription balance provides a buffer against gas volatility, and only the actual cost of your request will be deducted from your account.

>     If your subscription is underfunded, your VRF request will be pending for 24 hours. If that happens, check the [Subscription Manager](https://vrf.chain.link]) to see the additional balance needed.

Reviewer: ```suggestion
    If your subscription is underfunded, your VRF request will be pending for 24 hours. If this happens, check the [Subscription Manager](https://vrf.chain.link]) to see the additional balance needed.
```
